### PR TITLE
[core] Add native functions for UUIDv5 generation

### PR DESCRIPTION
**Pull request by behenate**
# Why

We want to generate UUIDv5 natively and be able to use that functionality in JS without importing additional modules.

# How

Created native implementations of the UUIDv5 algorithm (platforms don't provide it by default) and exposed them to JS with to `expo-modules-core`

# Test Plan

Tested in BareExpo on Android, iOS and web. UUIDs are consistent on all platforms for the same arguments.

### CODE REVIEW

File: packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import Foundation
+import CommonCrypto
+import CryptoKit
+func uuidv5(name: String, namespace: UUID) -> UUID {
```
Comment (lukmccall) [inline packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift:4]:
```suggestion

func uuidv5(name: String, namespace: UUID) -> UUID {
```

File: packages/expo-modules-core/CHANGELOG.md [hunk 2]
```diff
@@ -25,6 +25,8 @@
 - Fixed random `NullPointerExceptions` when calling `Updates.reloadAsync` on Android. ([#24442](https://github.com/expo/expo/pull/24442) by [@lukmccall](https://github.com/lukmccall))
 - Converting shared object id passed as a prop to the associated native object. ([#24431](https://github.com/expo/expo/pull/24431) by [@tsapeta](https://github.com/tsapeta))
 - Added a convenient method to get the app delegate subscriber of the given type. ([#23943](https://github.com/expo/expo/pull/23943) by [@tsapeta](https://github.com/tsapeta))
+- Use native UUIDv4 generation on `Android` and `iOS`. ([#24199](https://github.com/expo/expo/pull/24199) by [@behenate](https://github.com/behenate))
+- Use native UUIDv5 generation on `Android` and `iOS`. ([#24200](https://github.com/expo/expo/pull/24200) by [@behenate](https://github.com/behenate))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:29]: These changelog entries should go a bit up, right now they are under released 1.8.0 version 😉

File: packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import Foundation
+import CommonCrypto
+import CryptoKit
+
+func uuidv5(name: String, namespace: UUID) -> UUID {
+  var spaceUID = namespace.uuid
+  var data = withUnsafePointer(to: &spaceUID) { [count =  MemoryLayout.size(ofValue: spaceUID)] in
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift:7]:
```suggestion
  var data = withUnsafePointer(to: &spaceUID) { [count = MemoryLayout.size(ofValue: spaceUID)] in
```

File: packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  // Create a tuple for the final UUID
+  var uuidTuple = namespace.uuid
+
+  withUnsafeMutablePointer(to: &uuidTuple) { pointer in
+    let bound = pointer.withMemoryRebound(to: UInt8.self, capacity: 16) { $0 }
+    digest.enumerated().forEach { (bound + $0.offset).pointee = $0.element }
+  }
+
+  // Convert digest to UUID and return
+  return UUID.init(uuid: uuidTuple)
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift:31]:
```suggestion
  return UUID(uuid: uuidTuple)
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/defaultmodules/uuid/Uuidv5.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package expo.modules.kotlin.defaultmodules.uuid
```
Comment (tsapeta) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/defaultmodules/uuid/Uuidv5.kt:1]: I guess the 'defaultmodules' folder was intended to be just for modules. What's the best place for such helpers @lukmccall ?

Comment (lukmccall) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/defaultmodules/uuid/Uuidv5.kt:1]: I think we can move it to `expo/modules/kotlin/uuid5/*` instead.

Comment (tsapeta) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/defaultmodules/uuid/Uuidv5.kt:1]: * `expo/modules/kotlin/uuidv5/*`

File: packages/expo-modules-core/CHANGELOG.md (hunk 2 above)
Comment (behenate) [reply packages/expo-modules-core/CHANGELOG.md:29]: Oops, must've happened when resolving a conflict 😅

File: packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift (hunk 1 above)
Comment (behenate) [reply packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift:5]: @lukmccall Hmm this is weird, I commited code that contains this line (the blank space). It shows up in my local code and in the github web editor (press dot on the keyboard to open). I'm not sure why it's not visible in here 🧐🤔

File: packages/expo-modules-core/CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,9 @@
 
 ### 💡 Others
 
+- Use native UUIDv4 generation on `Android` and `iOS`. ([#24199](https://github.com/expo/expo/pull/24199) by [@behenate](https://github.com/behenate))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:13]: UUID v4 was already added in the previous version
```suggestion
```

File: packages/expo-modules-core/CHANGELOG.md [hunk 7]
```diff
@@ -49,7 +52,6 @@
 
 ### 💡 Others
 
-- Use native UUIDv4 generation on `Android` and `iOS`. ([#24199](https://github.com/expo/expo/pull/24199) by [@behenate](https://github.com/behenate))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:52]: Revert this change

File: packages/expo-modules-core/src/uuid/uuid.ts [hunk 8]
```diff
@@ -14,8 +14,25 @@ function uuidv4(): string {
   return nativeUuidv4();
 }
 
+function uuidv5(name: string, namespace: string | number[]) {
+  const parsedNamespace =
+    Array.isArray(namespace) && namespace.length === 16 ? bytesToUuid(namespace) : namespace;
+
+  // If parsed namespace is still an array it means that it wasn't valid
+  if (Array.isArray(parsedNamespace)) {
+    throw new Error('`namespace` must be uuid string or an Array of 16 byte values');
```
Comment (tsapeta) [inline packages/expo-modules-core/src/uuid/uuid.ts:23]: We should use consistent casing in error messages (`uuid` -> `UUID`)
```suggestion
    throw new Error('`namespace` must be a valid UUID string or an Array of 16 byte values');
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/modules/uuidv5/Exceptions.kt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package expo.modules.kotlin.modules.uuidv5
```
Comment (tsapeta) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/modules/uuidv5/Exceptions.kt:1]: See https://github.com/expo/expo/pull/24200#discussion_r1334153651

File: packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import Foundation
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Swift/Modules/Uuid/Uuidv5.swift:1]: Like on Android, let's move this file out of the `Modules` directory (`Uuidv5.swift` is not a module). It could be `ios/Swift/UUIDv5.swift`

**Review by tsapeta — APPROVED** [review]
